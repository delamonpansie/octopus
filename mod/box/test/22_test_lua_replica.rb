#!/usr/bin/ruby

$: << File.dirname($0) + '/lib'
require 'run_env'

class RunEnvX < RunEnv
  task :setup => "box_init.lua"
  file "box_init.lua" do
    File.open("box_init.lua", "w") do |io|
      io.write <<-EOD
        local box = require 'box'
        user_proc.select = box.wrap(function (shard, n)
	  local object_space = shard:object_space(n)
	  local pk = object_space:index(0)
	  return 0, {pk:find('0')}
        end)

        user_proc.update = box.wrap(function (shard, n, i)
          i = tonumber(i)
	  local object_space = shard:object_space(n)
	  local pk = object_space:index(0)
	  object_space:update('0', { i, 'delete' } )
	  return 0, {pk:find('0')}
        end)

	user_proc.error = box.wrap(function (shard, n)
          error('fooo')
        end)

        print('box_init.lua loadded')
      EOD
    end
  end

end

class MasterEnv < RunEnvX
  def test_root
    super << "_master"
  end

  def config
    super + <<EOD
wal_feeder_bind_addr = ":33034"
EOD
  end
end

class SlaveEnv < RunEnvX
  def initialize
    @primary_port = 33023
    @test_root_suffix = "_slave"
    super
  end

  def config
    super + <<EOD
wal_feeder_addr = "127.0.0.1:33034"
EOD
  end
end

master = MasterEnv.new.connect_eval do
  self.connect_name = "master"
  ping
  self
end

SlaveEnv.connect_eval do
  self.connect_name = "slave"
  master.insert ['0', 'a', 'b', 'c', 'd']
  wait_for { select_nolog(['0']).length > 0 }

  master.select ['0']
  select ['0']

  master.lua 'user_proc.select', '0'
  lua 'user_proc.select', '0'

  master.lua 'user_proc.update', '0', '2'
  log_try { lua 'user_proc.update', '0', '3' }

  log_try { lua 'user_proc.error', '0' }
end

