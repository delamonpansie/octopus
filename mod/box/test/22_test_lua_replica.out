# master.ping()
:pong

# master.insert(["0", "a", "b", "c", "d"])
1

# master.select(["0"])
[["0", "a", "b", "c", "d"]]

# slave.select(["0"])
[["0", "a", "b", "c", "d"]]

# master.lua("user_proc.select", "0")
[["0", "a", "b", "c", "d"]]

# slave.lua("user_proc.select", "0")
[["0", "a", "b", "c", "d"]]

# master.lua("user_proc.update", "0", "2")
[["0", "a", "c", "d"]]

# slave.lua("user_proc.update", "0", "3")
Failed with: {code: 0x102, message: 'txn is readonly
stack traceback:
	[C]: in function '_dispatch'
	mod/box/src-lua/box.lua:90: in function 'update'
	box_init.lua:12: in function 'proc'
	mod/box/src-lua/box.lua:389: in function <mod/box/src-lua/box.lua:381>'}
# slave.lua("user_proc.error", "0")
Failed with: {code: 0x202, message: 'box_init.lua:17: fooo
stack traceback:
	[C]: in function 'error'
	box_init.lua:17: in function 'proc'
	mod/box/src-lua/box.lua:389: in function <mod/box/src-lua/box.lua:381>'}
